### PR TITLE
RFC: Perf JIT event listener

### PR DESCRIPTION
**Pull request by maleadt**
Any interest in this? It's a premature version, but if people are interested I could look into getting it in a shape to merge.

This PR adds a JIT listener which dumps symbol information to `/tmp/perf-$PID.map`, which is the file `perf` reads when reporting. It makes profiling really easy on any somewhat-recent Linux version.

```
$ make USE_PERF_JITEVENTS=1 USE_ORCJIT=0  # this is not a real build var
$ ENABLE_JITPROFILING=1 \
  perf record --call-graph fp ./julia perf/shootout/pidigits.jl
$ perf report --call-graph -G
```

This yields the following call graph, zooming in on the main `pidigits` function :

```
  Children      Self  Comma  Shared Object   Symbol

+   12,66%     0,00%  julia  perf-12888.map  [.] julia_pidigits_23309
-   12,66%     0,00%  julia  perf-12888.map  [.] jlcall_pidigits_23309
     jlcall_pidigits_23309
   - julia_pidigits_23309
      - 12,60% julia_pidigits_23310
         - 4,64% julia_divrem_23315
            + 3,88% __gmpz_tdiv_qr
            + 0,61% jl_gc_alloc_2w
            + 0,08% jl_apply_generic
         + 2,74% julia_*_23313
         + 1,72% julia_+_23314
         + 1,41% julia_*_23319
         + 1,25% julia_<<_23312
         + 0,34% jl_apply_generic
         + 0,18% julia_-_23318
         + 0,14% __gmpz_init
           0,02% jl_gc_alloc_2w
           0,01% __gmpz_add
           0,01% __gmpz_mul_si
         + 0,01% julia_+_23314
        0,04% __gmpz_cmp
        0,01% julia_-_23318
        0,01% __gmpz_cmp
+   12,60%     0,12%  julia  perf-12888.map  [.] julia_pidigits_23310
```

Once the follow-up of [this patchset](https://lwn.net/Articles/638566/) lands it should even be possible to annotate source code, making this a really powerful tool. Time to introduce `deps/linux`? :grin:

### CODE REVIEW

**Comment by timholy** [comment]
Aside from the fact that the built-in profiler is currently broken on master (#14281), can you summarize the main advantages of `perf`? Presumably you can use it to get extra detail? (e.g., count the number of cache misses, etc?)

**Comment by maleadt** [comment]
The main motivation was to analyse how the compiler (including LLVM) interacts with user code, both performance-wise and just getting a nice call-stack. As far as I can see, the built-in profiler only shows Julia code. `perf` allows much more detailed analysis, probably only of interest to core developers.

Apart from that, there are some other advantages: overhead is much lower (on some random test, `perf` adds 10%, `@profile` adds 500%), there's the hardware details you mentioned, and `perf` is widely available (contrary to Intel VTune or maybe even OProfile).
On the other hand, it might not add much over the existing OProfile support, which is part of upstream LLVM.

**Comment by timholy** [comment]
If you pass `C=true` (to either `Profile.print` or the `ProfileView` package) you can get some of that information from the built-in profiler.

I've never seen `@profile` add 500%. Can you provide a concrete example? Here's one (on `julia 0.4.4-pre+2`):
```jl
julia> a = rand(10^7);

julia> @time sort(a);
  1.323654 seconds (8 allocations: 76.294 MB, 0.34% gc time)

julia> @time @profile sort(a);
  1.371522 seconds (4 allocations: 76.294 MB)
```
You can adjust the overhead with `Profile.init`.

Finally, the built-in profiler is even more widely available than `perf` :smile:.

I'm just making sure you know what's possible with the built-in profiler. If, after this, there's still stuff that `perf` can do that the built-in one can't, I'm all for adding support.

**Comment by maleadt** [comment]
You're right, those numbers are way off, no idea what happened (can't reproduce on my home system now). I'll report back tomorrow if I can reproduce.
And I (obviously) didn't know about `C=true`, thanks for mentioning! I prefer `perf`'s curses interface though, but that's a minor detail.

I guess it's just a compatibility debate then, whether we want to add support for yet another low-level JIT profiling interface, similar to the current ones (OProfile, VTune).

**Comment by timholy** [comment]
I'm just guessing at how it works, but the curses interface indeed looks nicer than a dump. However, for complicated situations you might like [ProfileView](https://github.com/timholy/ProfileView.jl).

Overall I have no objections to supporting perf, and this doesn't look particularly invasive. I'll shut up now and let folks who maintain `codegen.cpp` and the build system comment about how they feel about adding another hook.

**Comment by tkelman** [comment]
does it have to hardcode /tmp/ ? The data collection doesn't actually look platform dependent?

**Comment by maleadt** [comment]
It does [hardcode](https://github.com/torvalds/linux/blob/a200dcb34693084e56496960d855afdeaaf9578f/tools/perf/util/map.c#L169) `/tmp` indeed. Will change in the future, but not any time soon.
The data collection isn't really platform dependent, just hooks onto the LLVM JIT listener infrastructure and dumps symbol names and their corresponding addresses.

**Comment by maleadt** [comment]
Maybe it's also interesting to make it possible to enable the profiling functionality without having to recompile, ie. removing the `#ifdef`s, on the condition that it doesn't hurt performance?

**Comment by maleadt** [comment]
Ref https://groups.google.com/forum/#!topic/llvm-dev/yw809fWc20U

Although this PR is stale, I still think `perf` support is useful (eg. to inspect hardware performance counters or profile Julia start-up) so I'd leave this open as a tracking issue unless somebody objects.

**Comment by vchuravy** [comment]
There is now a PR open against LLVM to support PerfJITEvents: https://reviews.llvm.org/D44892

**Comment by DokFaust** [comment]
Hi @vchuravy! I've tried to patch

**Comment by DokFaust** [comment]
Hi @vchuravy! I have tried [to naively patch Julia](https://github.com/JuliaLang/julia/compare/master...DokFaust:df/perf-D44892) with the aforementioned LLVM review, but honestly it is problematic.

Indepently of Julia, compiling it on LLVM (__trunk__) and analyzing a JITted program as in the example can be quite confusing, as **lli** seems to be not developed to reach peak perfomance, so even the most resources-avid routine will consume a negligible share of CPU cycles ( < 1%).
This makes also reading line-level annotations a quite painful experience, but they seem to be annoted properly at least.

The reliance on **lli** seems an huge limitation to me, as using it on Julia requires a stand-alone sysimage in LLVM IR.  I can not see a way to avoid this, as even calling **lli** inside Julia would still present the same problems mentioned above.
Unfortunately I have not been able to test the patch with Julia as:

- The **lli** version that I had been able to compile with Julia seems broken : it is not even capable of printing an help message.

- After producing a Julia program bitcode, if I try to convert it to LLVM IR **llvm-dis** exits with error message ""Invalid bitcode signature"". Is _usr/lib/julia/basecompiler.ji_ the right sysimage to provide?

Anyway that's what came out of my tests ¯\_(ツ)_/¯

**Comment by vchuravy** [comment]
@DokFaust your patch does look about right, but you will also have to make sure to add `-DLLVM_USE_PERF=ON` to cmake in `llvm.mk`, similarly how we setup `LLVM_USE_INTEL_JITEVENTS`.

We should not need to use `lli`, but it should also not be broken. Once the `perf` support is compiled as part of Julia, Julia itself should emit the correct records, so that there is no need to use `lli`.

```
perf record -o /tmp/perf.data ./julia expensive_function.jl
perf inject --jit -i perf.data -o perf.jit.data
perf report -i perf.jit.data
```

> After producing a Julia program bitcode, if I try to convert it to LLVM IR llvm-dis exits with error message ""Invalid bitcode signature"". Is usr/lib/julia/basecompiler.ji the right sysimage to provide?
`basecompiler.ji` is not in bitcode format. There are two output options `--output-bc` and `--output-ji` the former produces LLVM bitcode the latter is used to create the sysimage.

**Comment by vchuravy** [comment]
Actually it turns out to be more complicated.

The support for *JITEVENTS (may it be perf/vtunes/oprofile) is dependent on MCJIT and Julia switched to OrcJIT completely in 0.7
So while the build system still has the remnants of the functionality I believe none of it is currently hooked up correctly.

cf: http://lists.llvm.org/pipermail/llvm-dev/2017-October/118134.html and #22471 and `NotifyGDB` in `jitlayers.cpp`

**Comment by vchuravy** [comment]
Closed in favour of #27466
